python/tvm/relay/frontend/change_datatype.py

```diff
@@ -0,0 +1,88 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+# pylint: disable=unused-argument
+"""Change Datatype Pass"""
+from ..function import Function
+from ..expr_functor import ExprMutator
+from ..transform.transform import function_pass
+from ..expr import var, bind
+
+# TODO(@gussmith23) what's the right opt level here?
+@function_pass(opt_level=0)
+class ChangeDatatype(ExprMutator):
+    """Mutator for changing the datatype of Relay programs.
+
+    Example:
+
+    .. code-block:: python
+
+        from tvm.relay.testing.inception_v3 import get_workload
+        expr, params = get_workload()
+
+        def change_dtype(src, dst, expr, params):
+            cdtype = ChangeDatatype(src, dst)
+            expr = cdtype.visit(expr)
+            expr = relay.ir_pass.infer_type(expr)
+            params = dict((p, tvm.nd.array(params[p].asnumpy().astype(dst))) for p in params)
+            return expr, params
+    """
+    def __init__(self, src, dst):
+        self.src = src
+        self.dst = dst
+        super().__init__()
+
+    def transform_function(self, func, mod, ctx):
+        return self.visit(func)
+
+    def visit_constant(self, const):
+        if const.data.dtype == self.src:
+            return const.astype(self.dst)
+        # TODO(hypercubestart): should we raise an error in this case, or return const?
+        return const
```

Comment:
@gussmith23 i added this here to avoid the linting, but I think casting it in every case makes more sense. What do you think?

Reply:
As I understand it, we may visit `const`s that are of different types -- types other than `self.src`, which is the type we're converting. If that's the case, we should leave those types alone. Check my logic on that though!

Reply:
I'm having trouble thinking of a case where const.data.dtype != src. In our tests, the only test that uses relay.ConstantNode is test_batch_norm where there is an epsilon constant and a 1f const, but the type of these constants is always the same as the type of src (maybe due to type inference?)

Reply:
We had a discussion about this:
Andrew's thoughts were twofold:
1. Are there ever any other datatypes other than `src` in workloads?
  - Good question, and I don't truly know, but I think it's better to assume that there are. For example, complicated workloads w/ control flow probably use counters and other `int`s. We don't want to mess with those!
2. With the current API, users could forget to convert some of their datatypes.
  - Totally true. My thought is, if a user has e.g. `float16` and `float32` in their workload, then
     1. They probably know that there are multiple datatypes being used in their workload, and should know that they have to convert both types, and
     2. if they forget to convert both types, BYODT should give them errors along the lines of "can't find lowering function for posit32 -> float16 cast", at which point they should hopefully realize the problem.